cbor/_common_encode.ts

```diff
@@ -1,188 +1,250 @@
-    const output = new Uint8Array(9);
+export function encode(
+  input: CborType,
+  output: Uint8Array,
+  offset: number,
+): number {
+  switch (typeof input) {
+    case "undefined":
+      output[offset++] = 0b111_10111;
+      break;
+    case "boolean":
+      output[offset++] = input ? 0b111_10101 : 0b111_10100;
+      break;
+    case "number":
+      return encodeNumber(input, output, offset);
+    case "bigint":
+      return encodeBigInt(input, output, offset);
+    case "string":
+      return encodeString(input, output, offset);
+    default:
+      if (input === null) output[offset++] = 0b111_10110;
+      else if (input instanceof Uint8Array) {
+        return encodeUint8Array(input, output, offset);
+      } else if (input instanceof Date) {
+        return encodeDate(input, output, offset);
+      } else if (input instanceof CborTag) {
+        return encodeTag(input, output, offset);
+      } else if (input instanceof Map) return encodeMap(input, output, offset);
+      else if (input instanceof Array) {
+        return encodeArray(input, output, offset);
+      } else return encodeObject(input, output, offset);
+  }
+  return offset;
+}
+
+function encodeHeader(
+  majorType: number,
+  input: number | bigint,
+  output: Uint8Array,
+  offset: number,
+): number {
+  if (input < 24) output[offset++] = majorType + Number(input);
+  else if (input < 2 ** 8) {
+    output[offset++] = majorType + 0b000_11000;
+    output[offset++] = Number(input);
+  } else {
     const view = new DataView(output.buffer);
-    if (x < 2 ** 16) {
-      output[0] = majorType + 25;
-      view.setUint16(1, x);
-      return output.subarray(0, 3);
-    }
-    if (x < 2 ** 32) {
-      output[0] = majorType + 26;
-      view.setUint32(1, x);
-      return output.subarray(0, 5);
+    if (input < 2 ** 16) {
+      output[offset++] = majorType + 0b000_11001;
+      view.setUint16(offset, Number(input));
+      offset += 2;
+    } else if (input < 2 ** 32) {
+      output[offset++] = majorType + 0b000_11010;
+      view.setUint32(offset, Number(input));
+      offset += 4;
+    } else {
+      output[offset++] = majorType + 0b000_11011;
+      view.setBigUint64(offset, BigInt(input));
+      offset += 8;
     }
-    if (x < 2 ** 64) {
-      output[0] = majorType + 27;
-      // Due to possible precision loss with numbers this large, it's best to do conversion under BigInt or end up with 1n off.
-      view.setBigUint64(1, BigInt(x));
-      return output;
+  }
+  return offset;
+}
+
+function encodeNumber(
+  input: number,
+  output: Uint8Array,
+  offset: number,
+): number {
+  if (input % 1 === 0) {
+    const isNegative = input < 0;
+    if (isNegative && input <= -(2 ** 64)) {
+      throw new RangeError(
+        `Cannot encode number: It (${input}) exceeds -(2 ** 64) - 1`,
+      );
+    } else if (input >= 2 ** 64) {
+      throw new RangeError(
+        `Cannot encode number: It (${input}) exceeds 2 ** 64 - 1`,
+      );
```

Comment:
IMO, these checks can lead to inaccurate results due to floating-point precision error

For example, `2 ** 64 - 10` is still equal to`2 ** 64`

Reply:
While precision is lost with numbers larger than 2 ** 53, I've experienced they're still consistent in how the number is interpreted and removing them would just cause the error to be thrown later at DataView.setBigUInt64 instead and would be less clear on why they got an error.